worker.go

```diff
@@ -22,23 +25,28 @@ const timerTypeName = "timer"
 
 // Worker is the doodad that does work.
 type Worker struct {
-	id          int
-	PacketChan  chan samplers.UDPMetric
-	ImportChan  chan []samplers.JSONMetric
-	QuitChan    chan struct{}
-	processed   int64
-	imported    int64
-	mutex       *sync.Mutex
-	traceClient *trace.Client
-	logger      *logrus.Logger
-	wm          WorkerMetrics
+	id               int
+	PacketChan       chan samplers.UDPMetric
+	ImportChan       chan []samplers.JSONMetric
+	ImportMetricChan chan *metricpb.Metric
+	QuitChan         chan struct{}
+	processed        int64
+	imported         int64
+	mutex            *sync.Mutex
+	traceClient      *trace.Client
+	logger           *logrus.Logger
+	wm               WorkerMetrics
 }
 
 // IngestUDP on a Worker feeds the metric into the worker's PacketChan.
 func (w *Worker) IngestUDP(metric samplers.UDPMetric) {
 	w.PacketChan <- metric
 }
 
+func (w *Worker) IngestMetric(m *metricpb.Metric) {
+	w.ImportMetricChan <- m
```

Comment:
A channel send here to pass along a single metric actually ends up being [a fairly high cost operation](http://jmoiron.net/blog/go-performance-tales) (This is something i'm looking at over with `IngestUDP()` as well, which is clearly the pattern you're following). My big concern would be that the attendant costs of send/receive on _every_ individual metric would end up being a prohibitively high cost.

to believe that we should be able to construct this API in a similar fashion.
Can you think of any ways - hacks are fine for now - to design this API such that we can pass batches of metrics? Given that the existing import channel on the worker has a channel that receives in batches, it seems like this ought to be feasible.

Reply:
That's a really interesting link, and obviously super similar to exactly this situation 😄 

I made this change in be2eceed8ebd0b27f973d4a07e1f1d5953b7763b by changing `importsrv.Server` to group metrics by their destination worker, and then sending them all at once over a single channel send operation.  Fortunately I think the changes are actually pretty straightforward and not particularly hacky at all!

Unfortunately this didn't seem to have a huge impact on performance from my benchmark of `(*importsrv.Server).SendMetrics` (also in this PR).  It does send the metrics over a channel, but the sent message goes nowhere.  Obviously this is quite different from how imports happen normally, but the benchmarks didn't show any real performance gain.  Just something interesting to note.